examples/stm32/nucleo-f429zi-freertos-mip/mcu.h

```diff
@@ -0,0 +1,161 @@
+  SETBITS(gpio->OSPEEDR, 3UL << (n * 2), ((uint32_t) speed) << (n * 2));
+}
+
+static inline void gpio_set_af(uint16_t pin, uint8_t af) {
+  GPIO_TypeDef *gpio = GPIO(PINBANK(pin));  // GPIO bank
+  int n = PINNO(pin);                       // Pin number
+  //gpio->AFR[n >> 3] &= ~(15UL << ((n & 7) * 4));
+  //gpio->AFR[n >> 3] |= ((uint32_t) af_num) << ((n & 7) * 4);
+  SETBITS(gpio->AFR[n >> 3], 15UL << ((n & 7) * 4),
+          ((uint32_t) af) << ((n & 7) * 4));
+}
+#endif
+
+static inline void gpio_write(uint16_t pin, bool val) {
+  GPIO_TypeDef *gpio = GPIO(PINBANK(pin));
+  gpio->BSRR |= (1U << PINNO(pin)) << (val ? 0 : 16);
+}
+
+static inline void gpio_toggle(uint16_t pin) {
+  GPIO_TypeDef *gpio = GPIO(PINBANK(pin));  // GPIO bank
+  uint32_t mask = BIT(PINNO(pin));
+  gpio->BSRR |= mask << (gpio->ODR & mask ? 16 : 0);
+}
+
+#define UART1 USART1
+#define UART2 USART2
+#define UART3 USART3
+
+static inline void uart_init(USART_TypeDef *uart, unsigned long baud) {
+  // https://www.st.com/resource/en/datasheet/stm32f429zi.pdf
+  uint8_t af = 0;           // Alternate function
+  uint16_t rx = 0, tx = 0;  // pins
+
+  if (uart == UART1) RCC->APB2ENR |= BIT(4);
+  if (uart == UART2) RCC->APB1ENR |= BIT(17);
+  if (uart == UART3) RCC->APB1ENR |= BIT(18);
+
+  if (uart == UART1) af = 4, tx = PIN('A', 9), rx = PIN('A', 10);
+  if (uart == UART2) af = 4, tx = PIN('A', 2), rx = PIN('A', 3);
+  if (uart == UART3) af = 7, tx = PIN('D', 8), rx = PIN('D', 9);
+
+  gpio_init(tx, GPIO_MODE_AF, GPIO_OTYPE_PUSH_PULL, GPIO_SPEED_HIGH, 0, af);
+  gpio_init(rx, GPIO_MODE_AF, GPIO_OTYPE_PUSH_PULL, GPIO_SPEED_HIGH, 0, af);
+  uart->CR1 = 0;                           // Disable this UART
+  uart->BRR = FREQ / APB2_PRE / baud;      // FREQ is a CPU frequency
+  uart->CR1 |= BIT(13) | BIT(2) | BIT(3);  // Set UE, RE, TE
+}
+
+static inline void uart_write_byte(USART_TypeDef *uart, uint8_t byte) {
+  uart->DR = byte;
+  while ((uart->SR & BIT(7)) == 0) spin(1);
+}
+
+static inline void uart_write_buf(USART_TypeDef *uart, char *buf, size_t len) {
+  while (len-- > 0) uart_write_byte(uart, *(uint8_t *) buf++);
+}
+
+static inline int uart_read_ready(USART_TypeDef *uart) {
+  return uart->SR & BIT(5);  // If RXNE bit is set, data is ready
+}
+
+static inline uint8_t uart_read_byte(USART_TypeDef *uart) {
+  return (uint8_t) (uart->DR & 255);
+}
+
+static inline bool timer_expired(uint32_t *t, uint32_t prd, uint32_t now) {
+  if (now + prd < *t) *t = 0;                    // Time wrapped? Reset timer
+  if (*t == 0) *t = now + prd;                   // Firt poll? Set expiration
+  if (*t > now) return false;                    // Not expired yet, return
+  *t = (now - *t) > prd ? now + prd : *t + prd;  // Next expiration time
+  return true;                                   // Expired, return true
+}
+
+static inline void clock_init(void) {                 // Set clock frequency
+  SCB->CPACR |= ((3UL << 10 * 2) | (3UL << 11 * 2));  // Enable FPU
```

Comment:
`DSB` + `ISB` is recommended after this to purge the pipeline, though I don't see any FP operations in the near future

Reply:
Yeah so I'd like to have a minimal code that does the task